pkg/mimirpb/errors.go

```diff
@@ -0,0 +1,29 @@
+// SPDX-License-Identifier: AGPL-3.0-only
+
+package mimirpb
+
+import (
+	"github.com/grafana/dskit/grpcutil"
+)
+
+// IsClientError returns true if err is a gRPC or HTTPgRPC error whose cause is a well known
+// client error.
+func IsClientError(err error) bool {
+	// This code is needed for backward compatibility.
+	if code := grpcutil.ErrorToStatusCode(err); code/100 == 4 {
+		return true
+	}
+
+	if status, ok := grpcutil.ErrorToStatus(err); ok {
+		for _, details := range status.Details() {
+			if errDetails, ok := details.(*ErrorDetails); ok {
+				return errDetails.Cause == BAD_DATA ||
+					errDetails.Cause == INGESTION_RATE_LIMITED ||
```

Comment:
Are we sure we want to mark `INGESTION_RATE_LIMITED` and `REQUEST_RATE_LIMITED` errors as client? If `distributor.service-overload-status-code-on-rate-limit-enabled` is set to true, the push handler maps these errors to status code `529`, and otherwise to status code `429`.
If status code `529` is returned, are we sure we want to consider them as client errors?

Reply:
I'm open to discuss it. BAD_DATA is what I really care in this PR, but added the rate limited too because until very recent times it was only a 429 and so was considered a client error by any check we do as `code / 100 == 4`.